File: Python/queue.py
class Cell:
  def __init__(self, value, next):
    super().__init__();
    self.__value = value;
    self.__next = next;
  def value(self):
    return self.__value;
  def next(self):
    return self.__next;
  def setValue(self, value):
    self.__value = value;
  def setNext(self, next):
    self.__next = next;
  def __str__(self):
    return "(" + self.__value.__str__() + "," + self.__next.__str__() + ")";

class Queue:
  def __init__(self):
    self.__first = None;
    self.__last = None;
    self.__size = 0;
  def add(self, value):
    if (self.__first == None):
      self.__first = self.__last = Cell(value, None);
    else:
      self.__last.setNext(Cell(value, None));
      self.__last = self.__last.next();
    self.__size = self.__size + 1;
  def first(self):
    assert (self.__first != None), {'Queue Empty'}
    return self.__first.value();
  def remove(self):
    assert (self.__first != None), {'Queue Empty'}
    firstvalue = self.__first.value();
    self.__first = self.__first.next();
    if (self.__first == None):
      self.__last = None;
    self.__size = self.__size - 1;
    return firstvalue;
  def size(self):
    return self.__size;
  def __str__(self):
    return "[" + self.toString(self.__first, True) + "]";
  def toString(self, current, first):
    if (current):
      return ("" if first else ",") + current.value().__str__() + self.toString(current.next(), False);
    else:
      return "";

q = Queue();
print(q);
print("Size: " + str(q.size()));
print();

q.add(1);
print(q);
print("First: " + str(q.first()));
print("Size: " + str(q.size()));
print();

q.add(2);
print(q);
print("First: " + str(q.first()));
print("Size: " + str(q.size()));
print();

q.add(3);
print(q);
print("First: " + str(q.first()));
print("Size: " + str(q.size()));
print();

print("Remove: " + str(q.remove()));
print(q);
print("First: " + str(q.first()));
print("Size: " + str(q.size()));
print();

print("Remove: " + str(q.remove()));
print(q);
print("First: " + str(q.first()));
print("Size: " + str(q.size()));
print();

print("Remove: " + str(q.remove()));
print(q);
print("Size: " + str(q.size()));

File: Ruby/queue.rb
class Cell
  :@m_value
  :@m_next

  def initialize value, n
    @m_value = value
    @m_next = n
  end

  def value
    @m_value
  end

  def next
    @m_next
  end

  def setNext n
    @m_next = n
  end

  public def to_s
    @m_value
  end
end

class Queue
  @m_first
  @m_last
  @m_size

  def initialize
    @m_first = nil
    @m_last = nil
    @m_size = 0
  end

  def add value
    if @m_last != nil
      @m_last.setNext (Cell.new value, nil)
      @m_last = @m_last.next
    else
      @m_first = Cell.new value, nil
      @m_last = @m_first
    end
    @m_size = @m_size + 1
  end

  def first
    if @m_first != nil
      @m_first.value
    else
      "Queue Empty"
    end
  end

  def remove
    if @m_first != nil
      firstvalue = @m_first.value
      @m_first = @m_first.next
      @m_size = @m_size - 1
      firstvalue
    else
      "Queue Empty"
    end
  end

  def size
    @m_size
  end

  def to_s
    "[#{to_sX @m_first, true}]"
  end

  private def to_sX current, first
    if current != nil
      "#{first ? "" : ","}#{current.value}#{to_sX current.next, false}"
    else
      ""
    end
  end
end

s = Queue.new
puts s
puts "Size: #{s.size}"
puts

s.add 1
puts s
puts "first: #{s.first}"
puts "Size: #{s.size}"
puts

s.add 2
puts s
puts "first: #{s.first}"
puts "Size: #{s.size}"
puts

s.add 3
puts s
puts "first: #{s.first}"
puts "Size: #{s.size}"
puts

puts "remove: #{s.remove}"
puts s
puts "first: #{s.first}"
puts "Size: #{s.size}"
puts

puts "remove: #{s.remove}"
puts s
puts "first: #{s.first}"
puts "Size: #{s.size}"
puts

puts "remove: #{s.remove}"
puts s
puts "Size: #{s.size}"
puts

puts "first: #{s.first}"
puts "remove: #{s.remove}"
